core/src/main/java/org/elasticsearch/index/query/TemplateQueryBuilder.java

```diff
@@ -75,23 +80,57 @@ public TemplateQueryBuilder(String template, Map<String, Object> vars) {
      * */
     @Deprecated
     public TemplateQueryBuilder(String template, ScriptService.ScriptType templateType, Map<String, Object> vars) {
-        this.templateString = template;
-        this.vars = vars;
-        this.templateType = templateType;
+        this(new Template(template, templateType, null, null, vars));
     }
 
     @Override
     protected void doXContent(XContentBuilder builder, Params builderParams) throws IOException {
         builder.field(TemplateQueryBuilder.NAME);
-        if (template == null) {
-            new Template(templateString, templateType, null, null, this.vars).toXContent(builder, builderParams);
-        } else {
-            template.toXContent(builder, builderParams);
-        }
+        template.toXContent(builder, builderParams);
     }
 
     @Override
     public String getWriteableName() {
         return NAME;
     }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        BytesReference querySource = context.executeQueryTemplate(template);
+        try (XContentParser qSourceParser = XContentFactory.xContent(querySource).createParser(querySource)) {
+            final QueryShardContext contextCopy = new QueryShardContext(context.index(), context.indexQueryParserService());
+            contextCopy.reset(qSourceParser);
+            return contextCopy.parseContext().parseInnerQuery();
```

Comment:
I think I would try not to call deprecated methods, also I think we need to do something with the collected named queries, since they go to a different context. Same as WrapperQueryBuilder in the end:

```
QueryBuilder result = contextCopy.parseContext().parseInnerQueryBuilder();
context.combineNamedQueries(contextCopy);
return result.toQuery(context);
```